### PR TITLE
ZEP 4: Metadata Conventions

### PR DESCRIPTION
**Pull request by rabernat**
This ZEP describes how communities can standardize conventions around metadata and layout of Zarr data
using user-defined attributes in order to meet domain-specific application needs without changes to the
core data model and specification, and without specification extensions.

### CODE REVIEW

**Comment by jakirkham** [comment]
cc @briannapagan (in case this is of interest to you)

**Comment by rabernat** [comment]
This is directly relevant to the forthcoming geozarr work, so that's why I wanted to push it out in draft form

**Comment by jstriebel** [comment]
Awesome, thanks a lot! As mentioned in https://github.com/zarr-developers/zarr-specs/issues/169, this is also relevant for the issues https://github.com/zarr-developers/zarr-specs/issues/139 and https://github.com/zarr-developers/zarr-specs/pull/144.

I like that this is becoming a separate ZEP, it never occurred to me to separate this from ZEP 1.

**Comment by jbms** [comment]
The current proposal just allows a group/array to have a single convention.  Perhaps for some use cases that makes sense.  But the example in the proposal is "units", which could easily interoperate with numerous other possible conventions.  Instead the naming of attributes could be done to allow multiple conventions to be used at once, for example:

```json
{"zarr_convention": {"units-v1": {"units": "m^2"}, ...}
```

or 

```json
{"units-v1": {"units": "m^2"}, ...}
```

or

```json
{"units-v1": "m^2", ...}
```

**Comment by rabernat** [comment]
Very good point Jeremy.

TBH, I'm on the fence about whether the convention even needs to be explicitly identified. Like, maybe it could be enough to say

> Arrays with the `units` attribute set are assumed to be using this convention

Of the proposals above, I definitely favor the first one because it doesn't touch the name of the actual attribute. I could also image

```
{
    "zarr_convention_units-v1": True,
    "zarr_convention_foobar-v2": True
}
```

**Comment by martindurant** [comment]
Strongly support this concept.

Question:
you mention the currently uncodified (by zarr) conventions in the wild. Is there something to be done about conventions that arise organically and are not described in the zarr docs?

**Comment by rabernat** [comment]
> Is there something to be done about conventions that arise organically and are not described in the zarr docs?

I think that it's natural for conventions to arise organically. Once there is sufficient alignment and adoption, they can be proposed as conventions.

For conventions created in the wild, or borrowed from other formats (e.g. CF Conventions), it could be hard to require the presence of the `zarr_conventions` attribute. (I'm thinking about e.g. converting from NetCDF to Zarr.) There needs to be a way to simply document existing conventions, without prescribing new attributes to be present.

**Comment by martindurant** [comment]
>  it could be hard to require the presence

Recommended, but not required?

File: draft/ZEP0004.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+If a consensus cannot be reached, the steering council may still decide to publish the convention, accompanied by a
+disclaimer that it is not a consensus, and noting any objections that were raised during the discussion.
+
+It is also possible that multiple, competing conventions exist in the same domain. While not ideal, it's not up to
+the Zarr community to resolve such domain-specific debates.
+These conventions should still be documented in a central location, which hopefully helps move towards alignment.
+
+### Updating a Convention
+
+Conventions should be versioned using incremental integers, starting from 1.
```
Comment (joshmoore) [inline draft/ZEP0004.md:134]: For me, reading these two sentences leaves a certain tension. Can we clarify if the first is really a SHOULD or is it more a MAY?

File: draft/ZEP0004.md [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+By separating the conventions from the underlying storage container, we enable more modularity in a domain's software ecosystem.
+With conventions, Zarr can be substituted more easily for HDF5 with minimal changes to the rest of the software stack.
+
+There are already many conventions operating today in Zarr. Prominent examples include
+
+- [Open Microscopy Next Generation File Format](https://ngff.openmicroscopy.org/latest/#biblio-ome-zarr-py) (OME-NGFF) and its implementation in the [ome-zarr-py](https://ome-zarr.readthedocs.io/en/stable/?) package
```
Comment (joshmoore) [inline draft/ZEP0004.md:160]:
```suggestion
- [Open Microscopy Next Generation File Format](https://ngff.openmicroscopy.org/) (OME-NGFF) and its implementation in the [ome-zarr-py](https://ome-zarr.readthedocs.io/) package
```

File: draft/ZEP0004.md [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+To create data with this convention using zarr-python, one would write
+
+```python
+import zarr
+z = zarr.open(
+    'example_with_units.zarr', mode='w', shape=(10000, 10000), chunks=(1000, 1000), dtype='f4'
+)
+z.attrs['units'] = 'm^2' 
+z.attrs['zarr_convention'] = "units-v1"
```
Comment (joshmoore) [inline draft/ZEP0004.md:80]: For me, this is the most (or really only) potentially contentious line in the ZEP. Similar to the ZEP1 discussion last night, I could see discussing at least:

 - if `zarr_` is a generally special prefix or if `zarr_convention` is a special string
 - if it's possible to have more than one convention active in a zgroup/zarray at a time
 - whether other values (like URLs) are acceptable as the values of the convention string

**Review by joshmoore — COMMENTED** [review]
Thanks for this, @rabernat! I very much look forward to having the conventions list on zarr-specs. :+1:

File: draft/ZEP0004.md (hunk 3 above)
Comment (jstriebel) [reply draft/ZEP0004.md:80]: I'm also wondering if this is needed at all. Instead, I'd rather just specify that the field `units` has a metadata convention. Nobody is forced to follow this convention, and exisiting arrays with such a field would potentially follow it automatically.

**Comment by jstriebel** [comment]
I'd propose to remove the `zarr_convention` key, and simply have a document which defines metadata convention (+ the process to add them as laid out in this ZEP :heart:). The user attributes could follow the metadata conventions (or not^^), e.g.:
```json
{
  "units": "m^2",
  "writer": "zarr-python",
  "origin": [12300, 45600],
  "convention-key": "valuenotfollowingtheconvention",
  "some-other-key": "foo",
  …
}
```

I see the conventions as a good place to discuss & establish standards between implementations, not as a strict mechanism that must be enforced. Also, IMO it can't be enforced, having the `zarr_convention` key doesn't avoid misusages of such conventions  (e.g. other interpretations, using inches when only metric units are defined, bugs, …).

**Comment by ivirshup** [comment]
About specifying the convention: I think it's really quite useful to know which specific convention is being used and to version them.

For example, if two groups want to use the `units` field, how should I know how to interpret that? What if you want to update the convention?

In `anndata`, we [specify conventions](https://anndata.readthedocs.io/en/latest/fileformat-prose.html) for our data with an `encoding-type` and `encoding-version` field in `.attrs` (both in hdf5 and zarr). We used to not, and it kinda sucked to figure out people IO errors or making any updates.

I'd also agree with the point made above that only allowing a single convention may be limiting. Maybe instead conventions could be stored like:

```
z.attrs["conventions"] = {"convention": "version", ...}
```

**Comment by joshmoore** [comment]
I tend to agree with Isaac. Maybe it helps to think through (and/or specify) what processing looks like. I've been debating whether to bring up my favorite soapbox (JSON-LD) as a way of specifying such metadata that already exists, has processing rules, etc. e.g.:

```
| Context                                          | Field    | Interpretation                |   |
|--------------------------------------------------|----------|-------------------------------|---|
| N/A                                              | units    | this-file#units               |   |
| {@context: {units: example.com/}}                | units    | example.com/units             |   |
| {@context: http://some-file.jsonld}              | units    | whatever-some-file-says/units |   |
| {@context: {ns: https://some-other-file.jsonld}} | ns:units | some-other-namespace/units    |   |
```

Obviously, there are a lot of different edge cases there, but I do like the idea of not building our own.

**Comment by rabernat** [comment]
From a practical point of view, it may be simply impossible to impose a hard requirement for convention identifiers. A big use case for us is transcoding NetCDF / HDF5 data that already exists into Zarr. This data was written 10+ years ago and the metadata is what it is.

I think way forward is for me to put together the template referred to above. This should have a section on "How to identify this convention".

**Comment by jbms** [comment]
> From a practical point of view, it may be simply impossible to impose a hard requirement for convention identifiers. A big use case for us is transcoding NetCDF / HDF5 data that already exists into Zarr. This data was written 10+ years ago and the metadata is what it is.

Is it not an option to convert the metadata at the same time as the data conversion happens?  It seems that during this data conversion is when you would have the most context for decoding any metadata.

**Comment by ivirshup** [comment]
> From a practical point of view, it may be simply impossible to impose a hard requirement for convention identifiers.

Totally fair. Without any sort of identifier or format requirement this seems to me like a listing of conventions used with zarr.

If this is the direction, I wonder if even the "consensus" requirement for new conventions could be softened to "noteworthy" or removed. If no namespace is being reserved, then I don't think the zarr team needs to take on the responsibility of figuring out if a field has consensus on a file format. Especially since it's so easy to break consensus. For instance, of the given examples aren't Xarray Zarr, GDAL, and GeoZarr competing?

I think there's definitely value in collecting lists of conventions building on top of zarr, and making that visible. However, I wonder if doing more than that (like establishing credentials based on consensus/ use) is something better left to standards repositories like [fairsharing.org](https://fairsharing.org/search?q=netcdf&taxonomies=Not%2520applicable)?

**Comment by rabernat** [comment]
Really good points @ivirshup. Yes, we definitely don't want to give ourselves (zarr developers) the job of mediating standards in different scientific domains. The only intention here is to provide a means to document an existing convention, not do any sort of evaluation or approval. I'll modify to reflect that.

**Comment by jbms** [comment]
On my end, I'm most interested in attributes that are relevant to general purpose tools like Neuroglancer, e.g. things like units, different types of labels.  If there is no way to unambiguously identify the metadata then it is much more complicated to make use of it.

**Comment by normanrz** [comment]
I think namespacing would be a good idea. In the OME-Zarr context, we are thinking about wrapping all the OME-specific metadata under the `ome` key in the attributes https://github.com/ome/ngff/issues/182. I think that would be useful to allow metadata from different metadata conventions to exist in the same group/array.

File: draft/ZEP0004.md [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+This ZEP itself describes the _process_ by which conventions may be proposed, discussed, accepted, and published by the Zarr community.
+This process is intended to be much more lightweight and informal than a spec change, which requires a ZEP.
+
+If this ZEP is accepted, Conventions will be added to the <https://zarr-specs.readthedocs.io/> website as a new top-level heading
+and a corresponding folder will be created in the `zarr-developers/zarr-specs` repo.
+A conventions template, which accompanies this ZEP, will be added to that folder.
+
+### New Convention Process
+
+New conventions are proposed via a pull-request to the `zarr-specs` repo which adds a new conventions document.
```
Comment (christophenoel) [inline draft/ZEP0004.md:117]: Not clear what the pull-request contains (a domain specific convention document ?). If it contains a convention document, a template should describe its structure ?

Comment (rabernat) [reply draft/ZEP0004.md:117]: Yes, a template is needed. I need to finish this PR.

**Comment by rabernat** [comment]
Thanks for everyone's patience, and apologies for being slow to finish up this draft.

I plan to prioritize this over the next few weeks. I'll respond to the comments above and push a new draft that incorporates the feedback.

**Comment by rabernat** [comment]
I have finally updated this ZEP. Thanks everyone for the patience. In my update, I incorporated the following changes

- Distinguish between a "legacy convention" (to acommodate existing data) and an "explicit convention"
- For explicit conventions `zarr_conventions` should be an array of strings, allowing multiple conventions to be composed together
- Addressed the possibility of namespacing
- Addressed the possibility of versioning

My goal here is to include the very good ideas that have been proposed in the discussion above as recommended best practices while retaining the ability to support legacy conventions and practices already in use in the community.

File: draft/ZEP0004.md [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+Multiple conventions may be present.
+
+For example, a group metadata JSON document with conventions present might look like this
+
+```
+{
+    "zarr_format": 3,
+    "node_type": "group",
+    "attributes": {
+        "zarr_conventions": ["units-v1", "foo],
```
Comment (normanrz) [inline draft/ZEP0004.md:140]:
```suggestion
        "zarr_conventions": ["units-v1", "foo"],
```

**Review by MSanKeys963 — APPROVED** [review]

**Comment by ivirshup** [comment]
@MSanKeys963, did a discussion get opened for this?

**Comment by rabernat** [comment]
Hi Isaac! I believe it's on me to open a PR where the discussion will happen. I will try to do that today.

**Comment by ivirshup** [comment]
👍

Is it not meant to be a Discussion (as opposed to a PR/ Issue)? I think this kind of discussion heavily benefits from threading.

Maybe this changed: https://github.com/zarr-developers/zeps/pull/27 ?

**Comment by rabernat** [comment]
TBH I think the ZEP process still has a lot of details to be ironed out. I agree a discussion makes sense.

**Comment by tasansal** [comment]
This idea is excellent; I would love to help push this forward. What is the best way to collaborate?

**Comment by rabernat** [comment]
@tasansal - the discussion is continuing in https://github.com/zarr-developers/zarr-specs/pull/262

The best way to collaborate would be to share your use cases there.

**Pull request merged**